Xamarin.Forms.Core/IndicatorView.cs

```diff
@@ -0,0 +1,256 @@
+		}
+
+		public bool HidesForSingleIndicator
+		{
+			get => (bool)GetValue(HidesForSingleIndicatorProperty);
+			set => SetValue(HidesForSingleIndicatorProperty, value);
+		}
+
+		public Color IndicatorColor
+		{
+			get => (Color)GetValue(IndicatorColorProperty);
+			set => SetValue(IndicatorColorProperty, value);
+		}
+
+		public Color SelectedIndicatorColor
+		{
+			get => (Color)GetValue(SelectedIndicatorColorProperty);
+			set => SetValue(SelectedIndicatorColorProperty, value);
+		}
+
+		public double IndicatorSize
+		{
+			get => (double)GetValue(IndicatorSizeProperty);
+			set => SetValue(IndicatorSizeProperty, value);
+		}
+
+		public IEnumerable ItemsSource
+		{
+			get => (IEnumerable)GetValue(ItemsSourceProperty);
+			set => SetValue(ItemsSourceProperty, value);
+		}
+
+		void ResetIndicatorStyles()
+		{
+			try
+			{
+				BatchBegin();
+				ResetIndicatorStylesNonBatch();
+			}
+			finally
+			{
+				BatchCommit();
+			}
+		}
+
+		void ResetIndicatorCount(int oldCount)
+		{
+			try
+			{
+				BatchBegin();
+				if (oldCount < 0)
+				{
+					oldCount = 0;
+				}
+
+				if (oldCount > Count)
+				{
+					RemoveRedundantIndicatorItems();
+					return;
+				}
+
+				AddExtraIndicatorItems();
+			}
+			finally
+			{
+				ResetIndicatorStylesNonBatch();
+				BatchCommit();
+			}
+		}
+
+		void ResetIndicatorStylesNonBatch()
+		{
+			foreach (var child in Items)
+			{
+				ApplyColor(child as View);
+			}
+
+			IndicatorLayout.IsVisible = Count > 1 || !HidesForSingleIndicator;
+		}
+
+		void ResetItemsSource(IEnumerable oldItemsSource)
+		{
+			if (oldItemsSource is INotifyCollectionChanged oldObservableCollection)
+			{
+				oldObservableCollection.CollectionChanged -= OnCollectionChanged;
+			}
+
+			if (ItemsSource is INotifyCollectionChanged observableCollection)
+			{
+				observableCollection.CollectionChanged += OnCollectionChanged;
+			}
+
+			OnCollectionChanged(ItemsSource, new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Reset));
+		}
+
+		void AddExtraIndicatorItems()
+		{
+			var oldCount = Items.Count;
+			for (var i = 0; i < Count - oldCount && i < MaximumVisibleCount - oldCount; ++i)
+			{
+				var size = IndicatorSize > 0 ? IndicatorSize : 10;
+				var indicator = IndicatorTemplate?.CreateContent() as View ?? new Frame
+				{
+					Padding = 0,
+					HasShadow = false,
+					BorderColor = Color.Transparent,
+					VerticalOptions = LayoutOptions.Center,
+					HorizontalOptions = LayoutOptions.Center,
+					WidthRequest = size,
+					HeightRequest = size,
+					CornerRadius = (float)size / 2
+				};
+				var itemTapGesture = new TapGestureRecognizer();
+				itemTapGesture.Tapped += (tapSender, tapArgs) => Position = Items.IndexOf(tapSender as View);
+				indicator.GestureRecognizers.Add(itemTapGesture);
+				Items.Add(indicator);
+			}
+		}
+
+		void RemoveRedundantIndicatorItems()
+		{
+			foreach (var item in Items.Where((v, i) => i >= Count).ToArray())
```

Comment:
I'd avoid LINQ here because it's difficult to read in its current form and there may be slight performance overhead in using it. You could create a for loop instead and remove the redundant items. Also, does it matter if you remove the items from the beginning or end of the item source?

Reply:
1) Well, i will handle it with while(Items.Count > Count) {  //remove }
2) No, it doesn't

Reply:
Yeah, avoid linq please

Reply:
done